Added an Edit Scene Variables Context Menu to the scenes in Project Manager

Somtozech opened the pull request:
Added an Edit Scene Properties to the Context Menu in the Project Manager Panel.

The Edit scene Properties pops up a modal for editing scene variables.

Edit scene properties on the context menu
![Screenshot (80)](https://user-images.githubusercontent.com/32250849/76566214-95e22600-64ac-11ea-9c6c-e229a4ab08b6.png)


Modal pops up when the menu item is clicked
![Screenshot (79)](https://user-images.githubusercontent.com/32250849/76566132-659a8780-64ac-11ea-84f0-e613bbbfe312.png)

closes #1505

PawBud commented:
Can you pls tell me what `.vscode/` does?

Somtozech commented:
@PawBud .vscode contains the workspace editor settings I used to work on the codebase( i set it to use eslint for formatting my code so as to obey eslint rules for the repo).

Somtozech commented:
For what the PR does. It opens the edit scene variables you see when you click the edit scene variables button in the component ScenePropertiesDialog

4ian commented on newIDE/app/src/ProjectManager/index.js:
```diff
@@ -738,6 +749,27 @@ export default class ProjectManager extends React.Component<Props, State> {
+          {!!this.state.layoutVariablesDialogOpen && (
+          <VariablesEditorDialog
+            open={this.state.layoutVariablesDialogOpen}
+            variablesContainer={project.getVariables()}
```
You're passing the project variables, so the the dialog will show the global variables :/

Somtozech replied:
@4ian Changed it to show the layout variables now. 
![Screenshot (82)](https://user-images.githubusercontent.com/32250849/76690504-d07cc780-6640-11ea-935b-ec63f355951f.png)

4ian commented:
Thanks for the PR!

> The Edit scene Properties pops up a modal for editing scene variables.

While this can be interesting, at this moment the button should be called... "Edit Scene Variables" :) 

Also, you're passing the project variables, meaning that you're showing the global variables.

=> you need to get more familiar and be more careful about what you implement - because the label is not properly indicating what is opened, and the content is not showing what you describe in the PR description.

PawBud commented:
> For what the PR does. It opens the edit scene variables you see when you click the edit scene variables button in the component ScenePropertiesDialog

I was trying to add the ScenePropertiesDialog to the context menu but nonetheless Kudos to you

Somtozech commented:
@4ian to get the current layout variables you have to be able to track the current layout and the functionality is not yet implemented in the Project Manager. Can you suggest a way I could go about this?

Somtozech commented:
@4ian I added the changes to clearly meet the description of my PR. 

Changed Edit Scene Properties -> Edit Scene Variables
![Screenshot (81)](https://user-images.githubusercontent.com/32250849/76690633-21d98680-6642-11ea-809e-0f0cc49a716d.png)

Set it to show layout variables instead of global variables
![Screenshot (82)](https://user-images.githubusercontent.com/32250849/76690639-374eb080-6642-11ea-9d02-04825bd3f981.png)

4ian commented on newIDE/app/src/ProjectManager/index.js:
```diff
@@ -282,6 +287,8 @@ export default class ProjectManager extends React.Component<Props, State> {
     variablesEditorOpen: false,
     extensionsSearchDialogOpen: false,
+    layoutVariablesDialogOpen: false,
+    layoutName: '',
```
layoutName is too generic: this is the name used for what?

4ian commented on newIDE/app/src/ProjectManager/index.js:
```diff
@@ -270,6 +274,7 @@ type State = {|
   variablesEditorOpen: boolean,
   extensionsSearchDialogOpen: boolean,
+  layoutVariablesDialogOpen: boolean,
 |};
```
The State you declare is not the same as what you use then

4ian commented on newIDE/app/src/ProjectManager/index.js:
```diff
@@ -299,6 +306,18 @@ export default class ProjectManager extends React.Component<Props, State> {
   }
 
+  _onEditLayoutVariable = (open: boolean, name: string) => {
+    this.setState({ layoutVariablesDialogOpen: open, layoutName: name });
```
You could use a single variable for both, if you make the string nullable :)

Somtozech replied:
Ok @4ian . So you suggesting I use a single variable to represent the layoutVariablesDialogOpen and layoutName?

4ian replied:
Yep

4ian commented on newIDE/app/src/ProjectManager/index.js:
```diff
@@ -738,6 +760,27 @@ export default class ProjectManager extends React.Component<Props, State> {
             }
           />
+          {!!this.state.layoutVariablesDialogOpen && (
+            <VariablesEditorDialog
```
Can we modify SceneEditor so that it's using this, rather than having the same implementation of VariablesEditorDialog used twice?

Somtozech replied:
The SceneEditor takes in other props that the project manager cannot provide at the moment (would have to rewrite the same implementation for variables/functions it takes as props).

4ian replied:
I mean SceneEditor has the same thing inside it:

https://github.com/4ian/GDevelop/blob/f1730c239dbaad98b3708d3e94d2586c597e08bb/newIDE/app/src/SceneEditor/index.js#L1235-L1238

This seems like an unecessary duplication of the this. How can you modify SceneEditor so that there is only one VariablesEditorDialog used for this specific case (i.e: the case of modifying scene variables)?

4ian commented on newIDE/app/src/ProjectManager/index.js:
```diff
@@ -738,6 +760,27 @@ export default class ProjectManager extends React.Component<Props, State> {
+            <VariablesEditorDialog
+              open={this.state.layoutVariablesDialogOpen}
+              variablesContainer={this.getLayoutVariables()}
+              onCancel={() => this._onEditLayoutVariable(false)}
```
Careful, if you don't pass something it will be undefined. Always pass all variables to your functions. JavaScript might allow you not to do it, but this will bite you very hard later.

Somtozech replied:
Can you explain better? I don't really understand.

4ian replied:
_onEditLayoutVariable is expecting 2 parameters in your code, but you're passing only one.

Somtozech replied:
Ok. I get it now. Thanks for the explanation :). Looking to add a default parameter to the _onEditLayoutVariable in cases any of the parameters are not defined. What do you think?

4ian replied:
I think we can all have in a single variable, using null to represents something not defined.

PawBud commented:
Is the issue solved @Somtozech ?

4ian commented:
@PawBud please don't comment to ask this, as you can see by yourself :) 
* There are still comments not addressed
* The PR is not merged
* The content of the PR is not addressing all that is needed in the linked issue.

This is creating noise for me and makes me lose time to check again PR that I've already checked ;)

PawBud commented:
@4ian I am extremely sorry that I wasted your time. I shall not repeat it

Somtozech commented:
@4ian I changed to a single variable and modified the SceneEditor to take in extra props so as to display the edit variables scene (solves the issue of duplicated code).

PawBud commented:
@Somtozech I think `Edit Scene Variables` should be in `GameSettings/SceneVariables`

4ian commented on newIDE/app/src/ProjectManager/index.js:
```diff
@@ -270,6 +275,7 @@ type State = {|
   projectPropertiesDialogOpen: boolean,
   variablesEditorOpen: boolean,
   extensionsSearchDialogOpen: boolean,
+  name: any,
```
That's a super generic name for a variable :/ 
Can you think of something more descriptive? Also never use any for types.

Somtozech replied:
ok. what of `currentLayoutName`?

4ian commented on newIDE/app/src/ProjectManager/index.js:
```diff
@@ -299,6 +306,19 @@ export default class ProjectManager extends React.Component<Props, State> {
     }
   }
 
+  _onEditLayoutVariable = (name: any = null) => {
```
Never use any, don't use default parameters

Somtozech replied:
I used any because the name variable could be of two types (null ie object and string)

4ian commented on newIDE/app/src/ProjectManager/index.js:
```diff
@@ -299,6 +306,19 @@ export default class ProjectManager extends React.Component<Props, State> {
+    const { name } = this.state;
+    if (!project || !project.hasLayoutNamed(name)) return null;
+
+    console.log(project.getLayout(name));
```
Extra console.log, remove them before commiting

4ian commented on newIDE/app/src/SceneEditor/index.js:
```diff
@@ -132,6 +133,7 @@ type Props = {|
   onChooseResource: ChooseResourceFunction,
   resourceExternalEditors: Array<ResourceExternalEditor>,
   isActive: boolean,
+  layoutVariablesDialogOpen: boolean,
```
Do you still need this?

Somtozech replied:
Yeah. It set's the current layoutVariablesDialogOpen in the SceneEditor state to use the layoutVariablesDialogOpen in the props as default value if layoutVariablesDialogOpen is passed in as props

4ian commented on newIDE/app/src/ProjectManager/index.js:
```diff
@@ -738,6 +761,20 @@ export default class ProjectManager extends React.Component<Props, State> {
             }
           />
+          {!!this.state.name && (
+            <SceneEditor
```
This is still duplicated in SceneEditor/index.js

Somtozech replied:
@4ian Changed it to use the Scene Editor directly. Don't really understand the current duplication you referring to ):. Could you a little more detail about what you meant?

4ian replied:
Oops sorry wrong line commented! :)

4ian replied:
I'm not sure actually to understand why you're putting the entire `SceneEditor` inside ProjectManager 😬
We for sure don't want to do that.

Somtozech replied:
I actually used `VariablesEditorDialog` in my first PR. But then you mentioned "duplication of code and modifying SceneEditor so that there is only one VariablesEditorDialog used for this specific case". I thought what you meant was modifying the `SceneEditor` to suit the use case for `ProjectManager` ):

4ian replied:
Ok I understand better :) 
So `VariablesEditorDialog` is the proper component to use. 

But we want to avoid having to have it both in ProjectManager *and* in SceneEditor if possible. While it's not a huge duplication, there is still some duplicated props like the messages to show on screen (`emptyExplanationMessage`, etc...).

Two possibilities:
* You create a SceneVariablesEditorDialog component, that pass the props that are common like the messages (`emptyExplanationMessage` and co) and just take a few less props. You then use it in ProjectManager and SceneEditor
* OR you put SceneVariablesEditorDialog in the ancestor of SceneEditor and ProjectManager which is... MainFrame. You add a a state to designate which scene (also called layout) has its variables edited. You then pass a prop to SceneEditor and ProjectManager that is a function to call to open the scene variables.

Advantage of first solution: we're not bloating MainFrame.
Advantage of second solution: we don't need a common `SceneVariablesEditorDialog`, because we only have a single place in the code where we use VariablesEditorDialog for a scene - so no duplication of `emptyExplanationMessage` and other props.

Usually I would prefer the second solution, but here because I'd like to avoid bloating too much MainFrame, maybe the first solution is better.

4ian commented:
> @Somtozech I think Edit Scene Variables should be in GameSettings/SceneVariables since that is where all the variables are present

No, scene variables are specific to each scene :)

Somtozech commented:
@4ian I added the SceneVariablesEditorDialog component for editing the scene variables and reused it in SceneEditor and ProjectManager

triptych commented:
I think  this PR is obsoleted by another merge: https://github.com/4ian/GDevelop/pull/1607

4ian commented:
Right the "solution" in the PR #1607 :)
